compiler/rustc_span/src/hygiene.rs

```diff
@@ -1408,58 +1399,9 @@ pub fn decode_syntax_context<D: Decoder, F: FnOnce(&mut D, u32) -> SyntaxContext
         return SyntaxContext::root();
     }
 
-    let pending_ctxt = {
-        let mut inner = context.inner.lock();
-
-        // Reminder: `HygieneDecodeContext` is per-crate, so there are no collisions between
-        // raw ids from different crate metadatas.
-        if let Some(ctxt) = inner.remapped_ctxts.get(raw_id as usize).copied().flatten() {
```

Comment:
Removing this cache was probably a bad idea.

Reply:
I remember that the mechanism here was designed to prevent redundant locking, though I'm not entirely sure if it relates to performance. Let's wait for the performance test results to confirm.

Reply:
#139228 showed almost the same regressions as https://github.com/rust-lang/rust/pull/129827#issuecomment-2764245187.

@bvanjoi could you resubmit this PR with `remapped_ctxts` caching restored and https://github.com/rust-lang/rust/pull/129827#discussion_r2024200819 addressed?